### PR TITLE
[Semi Auto] Refactor Replicated Rule

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
Function optimization

### PR changes
Others

### Description

Pcard-70448

- default_data_parallel rule : hack rule for support hybrid parallel quickly, should be mapped to a specific op on purpose.
- replicated rule: bottom line rule for all ops that have NO specific rule.
- Adapt Api for PHI format in dygraph.
- Adapt Api to distinguish argument of single tensor and argument of vector of single tensor in static graph.

### CODE REVIEW

File: paddle/fluid/pybind/auto_parallel_py.cc [hunk 1]
```diff
@@ -375,6 +376,44 @@ void BindAutoParallel(py::module *m) {
              }
              return self.InferForward(ctx);
            })
+      .def("infer_forward", /*for op that have vector argument*/
+           [](const phi::distributed::SpmdRule &self,
+              const std::vector<std::pair<int, int>> input_ranges,
```
Comment (chenwhql) [inline paddle/fluid/pybind/auto_parallel_py.cc]: 这里需要用引用吗

File: paddle/fluid/pybind/auto_parallel_py.cc [hunk 2]
```diff
@@ -375,6 +376,44 @@ void BindAutoParallel(py::module *m) {
...
+             to distingish between single tensor argument and vector argument of
+             one tensor: start - end == 0: single tensor start - end == 1:
+             vector containing one tensor input_ranges: [(0, 0), (1, 3), (3, 4)]
+             + input_specs: [t0, t1, t2, t3]  --> t0, [t1, t2], [t3]
+             */
+             phi::distributed::InferSpmdContext ctx;
+             paddle::small_vector<phi::distributed::DistMetaTensor,
+                                  phi::kInputSmallVectorSize>
+                 ins;
+             for (auto range : input_ranges) {
```
Comment (chenwhql) [inline paddle/fluid/pybind/auto_parallel_py.cc]: 这里是否也需要用引用避免拷贝

File: paddle/phi/infermeta/spmd_rules/default_data_parallel.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#include "paddle/phi/infermeta/spmd_rules/utils.h"
+
+namespace phi {
+namespace distributed {
+/**
+ * A **hack** rule with a strong assumtion that the first dimension of
+ * all the input and ouput tensors is the batch dimension (broadcast dimension),
+ * therefore, if any tensor's first dimension is sharded, the sharding would be
+ * propagating to all the other tensors (for tensor first dimension). All the
+ *other axes of tensors would be set as unshard (-1).
```
Comment (chenwhql) [inline paddle/phi/infermeta/spmd_rules/default_data_parallel.h]: 注释句首建议加个空格对齐

File: paddle/phi/infermeta/spmd_rules/replicated.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <vector>
+
+#include "paddle/phi/core/distributed/auto_parallel/dist_meta_tensor.h"
+#include "paddle/phi/core/distributed/type_defs.h"
+#include "paddle/phi/infermeta/spmd_rules/utils.h"
```
Comment (chenwhql) [inline paddle/phi/infermeta/spmd_rules/replicated.h:21]: utils.h是不是在.cc中include就可以了

**Review by chenwhql — DISMISSED** [review]
LGTM overall

File: paddle/fluid/pybind/auto_parallel_py.cc (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/pybind/auto_parallel_py.cc]: 修改了

File: paddle/fluid/pybind/auto_parallel_py.cc (hunk 2 above)
Comment (JZ-LIANG) [reply paddle/fluid/pybind/auto_parallel_py.cc]: 修改了， THX~

File: paddle/phi/infermeta/spmd_rules/replicated.h (hunk 4 above)
Comment (JZ-LIANG) [reply paddle/phi/infermeta/spmd_rules/replicated.h:21]: 不太行。因为下面phi 接口方法的定义中，需要把 helper template 实例化，所以必须include  helper template 的实现，无法通过 forward declaration 引入 helper。

File: paddle/phi/infermeta/spmd_rules/default_data_parallel.h (hunk 3 above)
Comment (JZ-LIANG) [reply paddle/phi/infermeta/spmd_rules/default_data_parallel.h]: done

File: paddle/fluid/pybind/auto_parallel_py.cc [hunk 5]
```diff
@@ -375,6 +376,44 @@ void BindAutoParallel(py::module *m) {
...
              return self.InferForward(ctx);
            })
+      .def("infer_forward",  // for op that have vector argument
+           [](const phi::distributed::SpmdRule &self,
+              const std::vector<std::pair<int, int>> &input_ranges,
+              const std::vector<DistTensorSpec> &input_specs,
+              const std::vector<phi::Attribute> &attrs) {
+             /*
+             to distingish between single tensor argument and vector argument of
+             one tensor: start - end == 0: single tensor start - end == 1:
```
Comment (zhiqiu) [inline paddle/fluid/pybind/auto_parallel_py.cc:386]: Commonly, the range is [start, end), i.e., start is inclusive and end is exclusive.  Maybe wo need another way to distingish between `tensor` and `[tensor]`,

File: paddle/phi/core/distributed/auto_parallel/inferspmd_utils.cc [hunk 6]
```diff
@@ -19,6 +19,17 @@ namespace distributed {
 
 void InferSpmdContext::EmplaceBackInput(DistMetaTensor input) {
   inputs_.emplace_back(std::move(input));
+  int index = static_cast<int>(inputs_.size());
```
Comment (zhiqiu) [inline paddle/phi/core/distributed/auto_parallel/inferspmd_utils.cc]: exchange line 21 and 22 to get right index?

Comment (JZ-LIANG) [reply paddle/phi/core/distributed/auto_parallel/inferspmd_utils.cc]: fixed, it is a bug, thx~

File: paddle/fluid/pybind/auto_parallel_py.cc (hunk 5 above)
Comment (JZ-LIANG) [reply paddle/fluid/pybind/auto_parallel_py.cc:386]: Yes，we plan to use PyObject* to deal with variadic arguments Python end and parser them in C++ end in Next PR

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by risemeup1 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**